feat(docs): KEP-2599: Decouple runtime lifecycle from TrainJobs to simplify updating runtimes (#3428)
kubeflow/training-operator: docs/proposals/2599-mutable-runtimes/README.md

* **Self-contained TrainJob**: once a TrainJob is created, its configuration is entirely self-contained. It only depends on itself or on resources it has created and owns. It does not depend on any external resources.
* **Remove finalizer on runtimes**: `TrainingRuntimes` and `ClusterTrainingRuntimes` should no longer need a finalizer.

### Non-Goals

* **Mutable TrainJobs**: we are not proposing any changes to the existing immutable fields of TrainJobs. These fields will remain immutable.

## Proposal

### User Stories

### Story 1

As a platform engineer, I want to be able to update or delete a training runtime without breaking any existing running or paused training jobs.

### Story 2

As a maintainer of Kubeflow Trainer, I want to be able to update or delete the default training runtimes included in a Kubeflow Trainer release without introducing breaking changes for users.

## Design details

We propose making the TrainJob only look up the runtime configuration on first reconciliation and instead store a "snapshot" of the runtime configuration in a separate object:

> * create a new namespaced custom resource `TrainingRuntimeSnapshot` with the same API as the `TrainingRuntime` resource. This is an internal resource and should only be created or updated by the trainer controller. Each `TrainJob` will have one `TrainingRuntimeSnapshot` with the same name and namespace as the `TrainJob`.

Reviewer: Using the same name as the TrainJob is clean, but what happens if a `TrainingRuntimeSnapshot` with that name already exists when the controller tries to create it?

Author: Good thought, though I think it shouldn't happen given the snapshot is internal and should only be created by this controller. For there to be a clash under normal operation, I guess there'd need to be two different TrainJobs with the same name and namespace? Could it happen if a trainjob is deleted and quickly recreated with the same name but different config?

Btw - this is consistent with the pattern used elsewhere, e.g. mpi secrets are named `<trainjob-name>-mpi-ssh-auth`.

* when a train job is reconciled, the controller first tries to fetch the `TrainingRuntimeSnapshot` for the job. If the snapshot does not exist, it looks up the `(Cluster)TrainingRuntime` referenced by the train job and creates a new `TrainingRuntimeSnapshot` resource. The snapshot resource has the same name and namespace as the `TrainJob`, and the same spec copied from the referenced `(Cluster)TrainingRuntime`.
* the `TrainJob` reconciliation logic gets the runtime configuration from the snapshot rather than the `(Cluster)TrainingRuntime`.
* the reconciliation is otherwise unchanged.
* the `TrainingRuntimeSnapshot` is automatically deleted when the train job is deleted using an `ownerReference` on the snapshot.

Additional changes:
* Remove the finalizer on the runtimes. It is no longer necessary as TrainJobs only need to reference runtimes on first reconciliation.
* Remove the `TrainJobWatcher` interface and associated implementations and boilerplate.

### API and RBAC changes

> The `TrainingRuntimeSnapshot` will have the following API:

Reviewer: Following up on recent issues with kueue : https://github.com/kubeflow/trainer/issues/2599#issuecomment-3947578705
The current TrainJob controller reconciles suspended jobs - it calls `reconcileObjects` for
any non-finished TrainJob . IIUC "Snapshot on first reconciliation" means a suspended TrainJob gets snapshotted immediately on creation, before
an external controller (e.g. Kueue integration) can patch `spec.runtimeRef` ?

Author: Yes this is a something I'm wanting to discuss in today's community call.

cc. @VassilisVassiliadis

Reviewer 2: AFAIK the runtimeRef field is currently immutable so I think we can break this down into different questions:

1. how do we decouple the lifecycle of a RuntimeTraining (RT)/ClusterRuntimeTraining (CRT) object from the TrainJob objects that reference them?
2. how do we allow external controllers (e.g. Kueue) or users to patch "important" fields in a TrainJob ?
3. do we want to allow mutating the runtimeRef field?

For example, if the goal is to allow starting a TrainJob using:
- all options from  a RT/CRT object, BUT
- with a different image

There's several ways to do this.
One is:
1. create a new RT/CRT object which is a copy of the old one but with an updated image
2. reference it

Another is:
1. Reference the "old" RT/CRT object,
2. override the image using the runtimePatches API (KEP-#2170 does not support mutating this field right now)

If we treat a RT/CRT as the "lowest priority blueprint" for a TrainJob and allowing the TrainJob itself, or even external controllers/users, to override the values of fields that exist in this "blueprint" then this KEP is a valid answer to the question:

> 1. how do we decouple the lifecycle of a RuntimeTraining (RT)/ClusterRuntimeTraining (CRT) object from the TrainJob objects that reference them?

Technically, we still need to lift the immutability of the `spec.runtimeRef` field so that it points to the Snapshot but we can do it in a way where the TrainJob webhook only allows changing the value from a RT/CRT to a RT/CRT-SnapShot this way it's immutable except during 1st reconciliation.

P.S. Some more context:

The above suggestion implies the following "blueprint" priority:
- highest priority: The runtimePatches api (this overrides everything else below)
- middle priority: the remaining fields of TrainJob (e.g. trainjob.spec.trainer, etc)
- lowest priority: the RT/CRT fields

Reviewer: +1 to @VassilisVassiliadis's framing of breaking this into three separate questions. For this KEP specifically, I think the cleanest path forward is: snapshot only when spec.suspend transitions from true to false (or on first reconciliation if already unsuspended). This gives external controllers the window they need to modify things while suspended, and the snapshot captures the final state. Would that work for both your use cases?

Reviewer 2: I think this should work.

Author: Thanks @VassilisVassiliadis — the breakdown into separate questions is very helpful.

For this KEP, I suggest we focus scope on the first question:

> how do we decouple the lifecycle of a RuntimeTraining (RT)/ClusterRuntimeTraining (CRT) object from the TrainJob objects that reference them?

The other two:

> how do we allow external controllers (e.g. Kueue) or users to patch "important" fields in a TrainJob?

> do we want to allow mutating the runtimeRef field?

seem broader in scope and likely need more dedicated design, given their impact on API semantics, maintainability, and testing. It may be better to address those in a follow-up KEP or directly as part of your kep.

Approaches like snapshot-on-unsuspend could be explored as part of that work, though we should also evaluate the runtimePatches api approach.

Importantly, I don’t think this proposal blocks those future directions - it preserves the current immutability behaviour but just decoupled the job from the runtime.

@VassilisVassiliadis if we're limited scope to the first question is there anything in the design that would impact the controller-driven workflow you’re thinking of or introduce constraints we should account for now?

Reviewer 2: @robert-bell I agree with everything you said. 

Especially:

> For this KEP, I suggest we focus scope on the first question: how do we decouple the lifecycle of a RuntimeTraining (RT)/ClusterRuntimeTraining (CRT) object from the TrainJob objects that reference them?

And:

> [the other 2 questions] seem broader in scope and likely need more dedicated design, given their impact on API semantics, maintainability, and testing. It may be better to address those in a follow-up KEP or directly as part of your kep.

I think this KEP is a valid solution to the "1st question" without introducing any blockers to finding answers for the other 2 questions.

The one thing that's hard to give my opinion on is the introduction of the TrainingRuntimeSnapshot CRD. That is something the core maintainers should comment on because they have a better view of Trainer than I do. My 2 cents is that this is fine because this proposed approach addresses an existing problem (lifecycles of RT/CRT tied to TrainJob objects) in a way that's better than the alternatives and since TrainingRuntimeSnapshot is an "internal" CRD users won't interact with it and thus they won't mind.

@andreyvelich what do you think ?

```go
// TrainingRuntimeSnapshot contains a point-in-time snapshot of a TrainingRuntime or ClusterTrainingRuntime as it was
// observed when a TrainJob was first reconciled.

> type TrainingRuntimeSnapshot struct {

Reviewer: Thinking can we add source annotation like 
  `trainer.kubeflow.org/snapshot-source: ClusterTrainingRuntime/pytorch-default` on the snapshot?
Without it, once a runtime is deleted, there's no way to know which runtime a TrainJob originally used. It may become useful for debugging, auditing, and the migration scenario where you need to verify what was captured.

Author: I did wonder this, but I wasn't sure it was necessary given the TrainJob runtimeRef is currently immutable. There's already lineage via the TrainJob. I also wasn't sure how useful this info is given the runtime could have changed between the resource creating and being deleted. Wdyt?

Reviewer: Fair point that runtimeRef provides lineage. I'd still lean towards adding it - it's one annotation and the value is in debugging production issues quickly. I think when you're using kubectl get trainingruntimesnapshots and trying to correlate with runtimes, having the source right on the snapshot saves a hop through the TrainJob. But not a hill I'll die on - fine to defer to implementation time if others feel the same. 👍

Author: It's a good point - I can see the benefit of the annotation, and it's low risk given it's only an annotation. I've updated the text.

	metav1.TypeMeta `json:",inline"`

	// metadata of the TrainingRuntimeSnapshot.
	// +optional
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// spec of the TrainingRuntimeSnapshot.
	// +optional
	Spec TrainingRuntimeSpec `json:"spec,omitempty,omitzero"`
}
```

The following additional RBAC permissions will be granted:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
rules:
- apiGroups:
  - trainer.kubeflow.org
  resources:
  - trainingruntimesnapshots
  verbs:
  - create
  - get
  - list
  - patch
  - update

>   - watch

Reviewer: nit: I guess`delete` verb is missing here. The controller might need it for edge cases? (e.g. snapshot recreation if ownerReference is wrong, or test cleanup).
The SSA `FieldOwner("trainer")` pattern used
elsewhere in the controller also benefits from having `delete` available.. WDYT?

Author: Good spot, though it looks like the other resources don't have `delete` permissions and we should be consistent. I guess we technically don't need delete because we allow the api server gc to delete the resources?

**Workaround creates proliferation:** The safer practice is creating new runtime objects for each update (e.g., `pytorch-2.0`, `pytorch-2.1`, `pytorch-2.1.1`) rather than modifying existing runtimes. However, this leads to a proliferation of nearly-identical runtimes that confuses users and creates maintenance burden.

Decoupling TrainJob behavior from runtime objects would also enable removing the finalizers currently used to protect runtimes from deletion while in use—finalizers that can create operational challenges like orphaned resources blocking namespace deletion.

**Note:** The original [Trainer v2 design](../2170-kubeflow-trainer-v2/README.md#the-training-runtime-api) anticipated runtimes being immutable with version control (see also [#2599](https://github.com/kubeflow/trainer/issues/2599)). However, a versioning mechanism has not been implemented. This KEP takes an alternative approach that makes runtimes fully mutable while preserving TrainJob immutability.

> 

Reviewer 2: I think this PR makes sense. 

As for the immutability at the level of CRT/RT I'm not sure that's actually necessary.

A CRT/RT is read by KFT during 1st reconciliation of a TrainJob and from that point onward the TrainJob object is no longer associated with the CRT/RT. Users wishing to inspect the CRT/RT that's tied to the TrainJob object can refer to the ConfigMap containing the CRT/RT snapshot. In the future, if there is a need to enforce immutability at the level of the snapshot, we can just replace the ConfigMaps with CRDs.

This way CRT/RT can be managed by platform administrators. I think that leaving CRTs/RTs mutable is in line with the motivation of this KEP i.e. "maintaining" a CRT/RT over time e.g. by updating to an image containing patches for security vulnerabilities etc.

That said, even if CRT/RT objects do become immutable one can just emulate "patching" them by deleting/recreating. This is almost safe because TrainJob objects will only reference a CRT/RT between their moment of creation and their 1st reconciliation.

# KEP-2599: Decouple runtime lifecycle from TrainJobs to simplify updating runtimes

## Authors
- Rob Bell (Red Hat)

> Assisted by Claude Code (Sonnet 4.5)

Reviewer 3: It is sufficient to add this to the commit message / PR description.

```suggestion
```

### Non-Goals

* **Mutable TrainJobs**: we are not proposing any changes to the existing immutable fields of TrainJobs. These fields will remain immutable.
* **Remove the existing implementation-level protections**: e.g. like not updating JobSets once they are created.
* **Remove finalizer on runtimes**: removing the finalizer from `TrainingRuntimes` and `ClusterTrainingRuntimes` is out of scope for this KEP. However, this KEP creates the foundation to enable finalizer removal in future work.

## Proposal

### User Stories

### Story 1

As a platform engineer, I want to be able to update or delete a training runtime without breaking any existing running or paused training jobs.

### Story 2

As a maintainer of Kubeflow Trainer, I want to be able to update or delete the default training runtimes included in a Kubeflow Trainer release without introducing breaking changes for users.

## Design details

We propose making the TrainJob only look up the runtime configuration on first reconciliation and store a "snapshot" of the runtime configuration in a configmap:

> * Create a `ConfigMap` to store the runtime snapshot. This is an internal resource and should only be created or updated by the trainer controller. Each `TrainJob` will have one ConfigMap named `{trainjob-name}-runtime-snapshot` in the same namespace as the `TrainJob`. The ConfigMap would be annotated with the source runtime it was copied from to aid debugging, e.g. `trainer.kubeflow.org/source-runtime: ClusterTrainingRuntime/<runtime-name>` or `trainer.kubeflow.org/source-runtime: TrainingRuntime/<runtime-namespace>/<runtime-name>`.

Reviewer 3: Do we really need this annotation?
Given that TrainJob has this information, and ConfigMap has ownerRef to it, users can always traceback to it.
Thoughts @tenzen-y @astefanutti @robert-bell?

Author: You're right, this isn't needed now the configmap has the entire runtime in it. I'll remove it.

> * When a train job is reconciled, the controller first tries to fetch the runtime snapshot ConfigMap for the job. If the snapshot does not exist, the controller looks up the `(Cluster)TrainingRuntime` referenced by the train job and creates a new ConfigMap containing a single data item with key `runtime` and value set to the yaml-formatted representation of the referenced `(Cluster)TrainingRuntime`.

Reviewer 3: ```suggestion
* When a TrainJob is reconciled, the controller first tries to fetch the runtime snapshot ConfigMap for the job. If the snapshot does not exist, the controller looks up the `(Cluster)TrainingRuntime` referenced by the train job and creates a new ConfigMap containing a single data item with key `runtime` and value set to the yaml-formatted representation of the referenced `(Cluster)TrainingRuntime`.
```

* The `TrainJob` reconciliation logic uses the runtime configuration contained in the ConfigMap snapshot rather than the current configuration in the `(Cluster)TrainingRuntime`.
* The reconciliation is otherwise unchanged.

> * The snapshot ConfigMap is automatically deleted when the train job is deleted using an `ownerReference` on the ConfigMap.

Reviewer 3: ```suggestion
* The snapshot ConfigMap is automatically deleted when the TrainJob is deleted using an `ownerReference` on the ConfigMap.
```

An example ConfigMap is:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: <trainjob-name>-runtime-snapshot
  ownerReferences:
    - apiVersion: trainer.kubeflow.org/v1alpha1
      kind: TrainJob
      name: <trainjob-name>
      uid: 12345678-1234-1234-1234-123456789abc
      controller: true
      blockOwnerDeletion: true
data:
  runtime: |
    apiVersion: trainer.kubeflow.org/v1alpha1
    kind: ClusterTrainingRuntime
    metadata:
      name: torch-distributed
      labels:
        trainer.kubeflow.org/framework: torch
    spec:
      mlPolicy:
        numNodes: 1
        torch: {}
      template:
        spec:
          replicatedJobs:
            - name: node
              template:
                metadata:
                  labels:
                    trainer.kubeflow.org/trainjob-ancestor-step: trainer
                spec:
                  template:
                    spec:
                      containers:
                        - name: node
                          image: pytorch/pytorch:2.10.0-cuda12.8-cudnn9-runtime
```

### API and RBAC changes

No new CRDs or RBAC changes are required.
### Migrations for existing resources

> **TrainJobs**: existing TrainJobs are automatically migrated on first reconciliation after upgrade. The controller creates a runtime snapshot ConfigMap for each non-finished TrainJob by copying the current state of its referenced runtime.

Reviewer 3: Would users expect that for already running TrainJobs we create ConfigMap? We can create it only for the newly created jobs.

Author: Yes, IMO existing TrainJobs should behave consistently with newly created ones. Treating them differently could be confusing for platform admins, e.g. runtimes would only be safe to mutate once all pre-existing jobs have finished.

Given the migration is automatic with no user impact I think we should keep behaviour consistent. I’m also proposing an integration test to verify the migration path.

Reviewer 3: @astefanutti @tenzen-y @VassilisVassiliadis @kaisoz Any thoughts on this?
Since we didn't introduce immutability into Runtime yet, the snapshot might be incorrect for already created TrainJobs.
However, that might be considered as bug, since Runtimes must be immutable according to the KEP.

Reviewer 3: Ref issue for Runtime mutability: https://github.com/kubeflow/trainer/issues/2599
After this KEP, we can relax rule for Runtime immutability, since snapshot will have the actual spec being used when TrainJob is created.

Reviewer 3: As we discussed in the call, we will move forward with snapshot creation on the TrainJob reconciliation loop.
